quinn-proto/src/connection/spaces.rs

```diff
@@ -476,6 +479,88 @@ impl PendingAcks {
     }
 }
 
+/// Helper for mitigating [optimistic ACK attacks]
+///
+/// A malicious peer could prompt the local application to begin a large data transfer, and then
+/// send ACKs without first waiting for data to be received. This could defeat congestion control,
+/// allowing the connection to consume disproportionate resources. We therefore occasionally skip
+/// packet numbers, and classify any ACK referencing a skipped packet number as a transport error.
+///
+/// Skipped packet numbers occur only in the application data space (where costly transfers might
+/// take place) and are distributed exponentially to reflect the reduced likelihood and impact of
+/// bad behavior from a peer that has been well-behaved for an extended period.
+///
+/// ACKs for packet numbers that have not yet been allocated are also a transport error, but an
+/// attacker with knowledge of the congestion control algorithm in use could time falsified ACKs to
+/// arrive after the packets they reference are sent.
+///
+/// [optimistic ACK attacks]: https://www.rfc-editor.org/rfc/rfc9000.html#name-optimistic-ack-attack
+pub(super) struct OptimisticAckDetector {
+    /// Next outgoing packet number to skip
+    next_skipped_packet_number: u64,
+    /// Most recently skipped packet number
+    prev_skipped_packet_number: Option<u64>,
+    /// Next packet number to skip is randomly selected from 2^n..2^n+1
+    exponent: u32,
+}
+
+impl OptimisticAckDetector {
+    pub(super) fn new(rng: &mut (impl Rng + ?Sized)) -> Self {
+        // First skipped PN is in 0..64
+        let exponent = 6;
+        Self {
+            next_skipped_packet_number: rng.gen_range(0..2u64.saturating_pow(exponent)),
+            prev_skipped_packet_number: None,
+            exponent,
+        }
+    }
+
+    pub(super) fn peek_packet_number(&self, space_id: SpaceId, space: &PacketSpace) -> u64 {
```

Comment:
So for the methods here, you've made the trade-off to hide `PacketSpace::get_tx_number()` from the `super` module in favor of passing in `SpaceId` here and checking it against `SpaceId::Data`. That gains us some compile-time safety from forcing calls to `get_tx_number()` to happen through `OpportunisticAckDetector`, but makes the code here a bit more complex due to having to pass in the `space_id`. Given that `PacketSpace::next_packet_number` is `pub(super)` anyway, maybe it isn't so bad to leave the decision to access it directly to callers from the `Connection` impl and avoid passing the argument?

Reply:
This was a break-even on LoC, but revealed that `peek` is only called in the data space anyway, and narrows interfaces, so I think it's a win :+1: